### ISSUE TITLE
pumvisible() is always false on TextChangedI

### ISSUE THREAD

**Issue by oblitum** [vim/vim#563]
Given the following `.vimrc`:

``` vim
function! FooBarComplete(findstart, base)
  if a:findstart
    return col('.') - 1
  else
    return ["Foo", "Bar"]
  endif
endfunction
set completefunc=FooBarComplete

function! OnTextChanged()
  if !pumvisible()
    echom getline('.')[col('.') - 2]
    call feedkeys( "\<C-X>\<C-U>\<C-P>", 'n' )
  endif
endfunction

au TextChangedI * call OnTextChanged()
```

The popup menu should be visible almost anytime in insert mode while typing,
still `echom getline('.')[col('.') - 2]`, which prints the character just typed, is printing
all characters typed while the popup is still visible, implying the `if !pumvisible()` test
is not working.

Is there any reason for that? Is there a way to do what the code is trying to do or
is it a Vim bug that needs to get fixed?

**Comment by oblitum** [comment]
I believe the following `.vimrc` is more explicit: 

``` vim
function! FooBarComplete(findstart, base)
  if a:findstart
    return col('.') - 1
  else
    return ["Foo", "Bar"]
  endif
endfunction
set completefunc=FooBarComplete

function! OnTextChanged()
  if pumvisible()
    echom "pum visible"
  else
    echom "pum not visible"
    call feedkeys( "\<C-X>\<C-U>\<C-P>", 'n' )
    if pumvisible()
      echom "pum got visible"
    else
      echom "pum still not visible"
    endif
  endif
endfunction

au TextChangedI * call OnTextChanged()
```

I only get "pum not visible" and "pum still not visible" while typing.

**Comment by oblitum** [comment]
s/pum is not visible/pum not visible/

**Comment by vim-ml** [comment]
Hi Francisco,

2016-1-17(Sun) 3:59:14 UTC+9 Francisco Lopes:

> s/pum is not visible/pum not visible/

I think this is expected behavior.

feedkeys() Only queuing string to typeahead-buffer.
It does not yet process of the specified string in this function.
If it is called from a function that was registered to the autocommand event, it is processed after return from the function.

Therefore, pumvisible() of the results will not change in the OnTextChanged().

In addition, TextChangedI autocommand event does not triggered when the popup menu is visible.  (Pleese see document. `:help TextChangedI`)

p.s.
getline ('') and col ('') are wrong argument.
Maybe, '.' is you wanted to specify?

## 

Best regards,
Hirohito Higashi (a.k.a h_east)

**Comment by oblitum** [comment]
Hi Hirohito, I have specified that, but I'm talking from github bugtracker (https://github.com/vim/vim/issues/563), so it may not be arriving the mailing list correctly.

Thanks for the info, I did read the docs before but I still got confused because while the popup is visible, TextChangedI is still triggered. The example code provides an always visible popup, while at the same time TextChangedI is always triggered.

I'm not sure whether I'm hiting some kind of "short-circuit" situation, where it gets ON and OFF.

**Comment by vim-ml** [comment]
Hi Francisco,

2016-1-18(Mon) 12:57:11 UTC+9 Francisco Lopes:

> Hi Hirohito, I have specified that, but I'm talking from github bugtracker (#563), so it may not be arriving the mailing list correctly.

Oh, sorry.  I just recognized.

> Thanks for the info, I did read the docs before but I still got confused because while the popup is visible TextChangedI is still triggered. The example code provides an always visible popup, while at the same time TextChangedI is always triggered.
> 
> I'm not sure whether I'm hiting some kind of "short-circuit" situation, where it gets ON and OFF.

Please tell me input keys that step to reproduce.

## 

Best regards,
Hirohito Higashi (a.k.a h_east)

**Comment by oblitum** [comment]
Hi Hiro, I'm gonna close this issue, I think it was a misunderstanding of me. I was trying to avoid call feedkeys in TextChangedI if popup menu was visible but I realized that if it was being shown on all call to feedkeys, there would be a tightloop, but there isn't one, so what should be happening is that when entering any key, popup vanishes but TextChangedI is triggered which pops it up again, which when I type again any key, it vanishes again, but TextChangedI is triggered again, which pops the menu again, etc.

**Comment by vim-ml** [comment]
Hi Francisco,

2016-1-18(Mon) 16:35:23 UTC+9 Francisco Lopes:

> Hi Hiro, I'm gonna close this issue, I think it was a misunderstanding of me. I was trying to avoid call feedkeys in TextChangedI if popup menu was visible but I realized that if it was being called on all call to feedkeys, there would be a tightloop, but there isn't one, so what should be happening is that when entering any key, popup vanishes but TextChangedI is triggered which pops it up again, which when I type again any key, it vanishes again, but TextChangedI is triggered again, which pops the menu again, etc.

I understand your report, maybe.

`pumvisible ()` does not mean return whether the completion mode.
Returns whether the "popup menu is visible".
Whether completion mode it is not related.

Did you become clear on this?

## 

Best regards,
Hirohito Higashi (a.k.a h_east)

**Comment by oblitum** [comment]
I'm not sure whether you have got what I intended. I did want to know whether "popup menu was visible" instead of knowning about completion mode.

I think my understanding now may be correct because my explanation applies. TextChangedI is not triggered when popup menu is visible, as it's stated in the docs, but typing make the popup go away, and it also triggers TextChangedI, which brings back the popup through `call feedkeys( "\<C-X>\<C-U>\<C-P>", 'n' )`, which doesn't trigger TextChangedI since popup gets visible.

Not sure whether that's totally correct but, anyway, I guess I'm OK with what I have now.
